tokio-quiche/src/quic/router/multipath.rs

```diff
@@ -0,0 +1,684 @@
+            peer_addr,
+            pending_cid: pending_cid.clone(),
+            with_gso: self.config.has_gso,
+            pacing_offload: self.config.pacing_offload,
+            with_pktinfo: if self.primary_local_addr.is_ipv4() {
+                self.config.has_ippktinfo
+            } else {
+                self.config.has_ipv6pktinfo
+            },
+        };
+
+        let handshake_info = crate::quic::connection::HandshakeInfo::new(
+            handshake_start_time,
+            self.config.handshake_timeout,
+        );
+
+        // Create a new InitialQuicConnection with multipath support
+        // We need to create a new struct to hold multipath information
+        let conn = InitialQuicConnection::new(
+            crate::quic::connection::QuicConnectionParams {
+                writer_cfg,
+                initial_pkt,
+                shutdown_tx: shutdown_tx.clone(),
+                conn_map_cmd_tx: self.conn_map_cmd_tx.clone(),
+                scid: scid.clone(),
+                metrics: self.metrics.clone(),
+                #[cfg(feature = "perf-quic-listener-metrics")]
+                init_rx_time,
+                handshake_info,
+                quiche_conn: conn,
+                sockets: self.path_txs.clone(),
+                local_addrs: self
+                    .path_rxs
+                    .iter()
+                    .map(|(_, addr)| *addr)
+                    .collect(),
+                peer_addr,
+                packet_scheduler: self.config.packet_scheduler.clone(),
+            },
+        );
+
+        conn.audit_log_stats.set_transport_handshake_start(
+            crate::quic::router::instant_to_system(handshake_start_time),
+        );
+
+        self.conns.insert(scid, &conn);
+
+        // Add the client-generated "pending" connection ID to the map as well.
+        if let Some(pending_cid) = pending_cid {
+            self.conns.map_cid(pending_cid, &conn);
+        }
+
+        self.metrics.accepted_initial_packet_count().inc();
+        if self.config.enable_expensive_packet_count_metrics {
+            if let Some(peer_ip) =
+                quic_expensive_metrics_ip_reduce(conn.peer_addr().ip())
+            {
+                self.metrics
+                    .expensive_accepted_initial_packet_count(peer_ip)
+                    .inc();
+            }
+        }
+
+        send_permit.send(Ok(conn));
+        Ok(())
+    }
+}
+
+impl<Tx, Rx, M, I> MultiPathInboundRouter<Tx, Rx, M, I>
+where
+    Tx: DatagramSocketSend + Send + Sync + 'static,
+    Rx: DatagramSocketRecv,
+    M: Metrics,
+    I: InitialPacketHandler,
+{
+    // Poll a single path for incoming data
+    fn poll_path(
+        &mut self, cx: &mut Context<'_>, path_id: usize,
+    ) -> Poll<io::Result<PollRecvData>> {
+        #[cfg(not(target_os = "linux"))]
+        {
+            // Simple polling for non-Linux platforms
+            let (ref mut rx, _local_addr) = &mut self.path_rxs[path_id];
```

Comment:
As previously noted, this may cause troubles if there are the amount of specified paths (e.g., 4), but with Path IDs greater or equal than the number of paths (e.g., 0, 1, 3, 4, with Path ID 2 that has been abandoned and Path ID 4 created later on).